File: scripts/data_loaders/python/mps_loader.py
"""
MPS File Loader for Linear Programming problems.

This module provides functionality to load and parse optimization problems
stored in MPS (Mathematical Programming System) format, a standard format
for linear programming problems commonly used by Netlib and other LP libraries.

MPS Format Structure:
- NAME: Problem name
- ROWS: Row (constraint) definitions with types:
  - N: Objective function (free row)
  - E: Equality constraint (=)
  - L: Less-than-or-equal constraint (<=)
  - G: Greater-than-or-equal constraint (>=)
- COLUMNS: Variable coefficients in constraints
- RHS: Right-hand side values
- BOUNDS: Variable bounds (optional)
- RANGES: Range constraints (optional)
- ENDATA: End of data

The problem format is: min c'x subject to Ax {<=,=,>=} b, l <= x <= u
"""

import sys
from pathlib import Path
from typing import Any, Dict

import numpy as np
from scipy.sparse import lil_matrix

# Add project root to path for imports
project_root = Path(__file__).parent.parent.parent.parent
sys.path.insert(0, str(project_root))

from scripts.data_loaders.problem_loader import ProblemData
from scripts.utils.logger import get_logger

logger = get_logger("mps_loader")


class MPSLoader:
    """Loader for optimization problems in MPS format."""

    def __init__(self):
        """Initialize the MPS loader."""
        pass

    def load(self, file_path: str, problem_name: str = None) -> ProblemData:
        """
        Load problem from .mps file.

        Args:
            file_path: Path to the .mps file
            problem_name: Optional problem name (if not provided, extracted from file)

        Returns:
            ProblemData object
        """
        if not Path(file_path).exists():
            raise FileNotFoundError(f"File not found: {file_path}")

        logger.info(f"Loading MPS file: {file_path}")

        # Parse MPS file
        mps_data = self._parse_mps_file(file_path)

        # Use provided name, MPS NAME field, or filename as fallback
        if problem_name is None:
            problem_name = mps_data.get("name", Path(file_path).stem)

        # Convert to ProblemData
        problem_data = self._convert_to_problem_data(mps_data, problem_name)

        logger.info(f"Successfully loaded MPS problem: {problem_data}")
        return problem_data

    def _parse_mps_file(self, file_path: str) -> Dict[str, Any]:
        """
        Parse MPS format file.

        Args:
            file_path: Path to MPS file

        Returns:
            Dictionary containing parsed MPS data
        """
        mps_data = {
            "name": "",
            "rows": {},  # row_name -> type (N, E, L, G)
            "row_order": [],  # ordered list of row names
            "columns": {},  # col_name -> {row_name: coef}
            "col_order": [],  # ordered list of column names
            "rhs": {},  # row_name -> value
            "bounds": {},  # col_name -> {'lo': val, 'up': val, 'type': type}
            "ranges": {},  # row_name -> value
            "obj_name": None,  # name of objective row
        }

        with open(file_path, "r") as f:
            lines = f.readlines()

        current_section = None

        for line in lines:
            # Skip empty lines and comments
            line = line.rstrip("\n\r")
            if not line or line.startswith("*"):
                continue

            # Check for section headers
            upper_line = line.upper().strip()
            if upper_line.startswith("NAME"):
                mps_data["name"] = line.split()[1] if len(line.split()) > 1 else ""
                current_section = "NAME"
                continue
            elif upper_line == "ROWS":
                current_section = "ROWS"
                continue
            elif upper_line == "COLUMNS":
                current_section = "COLUMNS"
                continue
            elif upper_line == "RHS":
                current_section = "RHS"
                continue
            elif upper_line == "BOUNDS":
                current_section = "BOUNDS"
                continue
            elif upper_line == "RANGES":
                current_section = "RANGES"
                continue
            elif upper_line == "ENDATA":
                break

            # Parse section content
            if current_section == "ROWS":
                self._parse_rows_line(line, mps_data)
            elif current_section == "COLUMNS":
                self._parse_columns_line(line, mps_data)
            elif current_section == "RHS":
                self._parse_rhs_line(line, mps_data)
            elif current_section == "BOUNDS":
                self._parse_bounds_line(line, mps_data)
            elif current_section == "RANGES":
                self._parse_ranges_line(line, mps_data)

        return mps_data

    def _parse_rows_line(self, line: str, mps_data: Dict[str, Any]):
        """Parse a line from ROWS section."""
        parts = line.split()
        if len(parts) >= 2:
            row_type = parts[0].upper()
            row_name = parts[1]
            mps_data["rows"][row_name] = row_type
            mps_data["row_order"].append(row_name)
            if row_type == "N" and mps_data["obj_name"] is None:
                mps_data["obj_name"] = row_name

    def _parse_columns_line(self, line: str, mps_data: Dict[str, Any]):
        """Parse a line from COLUMNS section."""
        parts = line.split()
        if len(parts) < 3:
            return

        col_name = parts[0]

        # Initialize column if not seen before
        if col_name not in mps_data["columns"]:
            mps_data["columns"][col_name] = {}
            mps_data["col_order"].append(col_name)

        # Parse pairs of (row_name, coefficient)
        i = 1
        while i < len(parts) - 1:
            row_name = parts[i]
            try:
                coef = float(parts[i + 1])
                mps_data["columns"][col_name][row_name] = coef
                i += 2
            except (ValueError, IndexError):
                break

    def _parse_rhs_line(self, line: str, mps_data: Dict[str, Any]):
        """Parse a line from RHS section."""
        parts = line.split()
        if len(parts) < 3:
            return

        # First field is RHS name (usually ignored), then pairs of (row, value)
        i = 1
        while i < len(parts) - 1:
            row_name = parts[i]
            try:
                value = float(parts[i + 1])
                mps_data["rhs"][row_name] = value
                i += 2
            except (ValueError, IndexError):
                break

    def _parse_bounds_line(self, line: str, mps_data: Dict[str, Any]):
        """Parse a line from BOUNDS section."""
        parts = line.split()
        if len(parts) < 3:
            return

        bound_type = parts[0].upper()
        # parts[1] is bound name (usually ignored)
        col_name = parts[2]

        if col_name not in mps_data["bounds"]:
            mps_data["bounds"][col_name] = {"lo": 0.0, "up": np.inf, "type": "default"}

        if bound_type == "LO":  # Lower bound
            mps_data["bounds"][col_name]["lo"] = float(parts[3])
        elif bound_type == "UP":  # Upper bound
            mps_data["bounds"][col_name]["up"] = float(parts[3])
        elif bound_type == "FX":  # Fixed value
            val = float(parts[3])
            mps_data["bounds"][col_name]["lo"] = val
            mps_data["bounds"][col_name]["up"] = val
            mps_data["bounds"][col_name]["type"] = "fixed"
        elif bound_type == "FR":  # Free variable
            mps_data["bounds"][col_name]["lo"] = -np.inf
            mps_data["bounds"][col_name]["up"] = np.inf
            mps_data["bounds"][col_name]["type"] = "free"
        elif bound_type == "MI":  # Minus infinity (lower bound)
            mps_data["bounds"][col_name]["lo"] = -np.inf
        elif bound_type == "PL":  # Plus infinity (upper bound)
            mps_data["bounds"][col_name]["up"] = np.inf
        elif bound_type == "BV":  # Binary variable
            mps_data["bounds"][col_name]["lo"] = 0.0
            mps_data["bounds"][col_name]["up"] = 1.0
            mps_data["bounds"][col_name]["type"] = "binary"

    def _parse_ranges_line(self, line: str, mps_data: Dict[str, Any]):
        """Parse a line from RANGES section."""
        parts = line.split()
        if len(parts) < 3:
            return

        # First field is range name (usually ignored), then pairs of (row, value)
        i = 1
        while i < len(parts) - 1:
            row_name = parts[i]
            try:
                value = float(parts[i + 1])
                mps_data["ranges"][row_name] = value
                i += 2
            except (ValueError, IndexError):
                break

    def _convert_to_problem_data(self, mps_data: Dict[str, Any], problem_name: str) -> ProblemData:
        """
        Convert parsed MPS data to unified SeDuMi-style ProblemData format.

        This converts LP problems to SeDuMi standard form:
            min  c'x
            s.t. A_eq @ x = b_eq
                 x in K (cone constraints)

        All inequality constraints are converted to equality constraints with
        slack variables, and variable bounds are handled through the cone structure.

        Args:
            mps_data: Parsed MPS data
            problem_name: Name for the problem

        Returns:
            ProblemData object in SeDuMi format
        """
        n_orig_cols = len(mps_data["col_order"])

        # Build column index mapping
        col_idx = {name: i for i, name in enumerate(mps_data["col_order"])}

        # Separate rows by type
        obj_rows = []
        eq_rows = []
        le_rows = []
        ge_rows = []

        for row_name in mps_data["row_order"]:
            row_type = mps_data["rows"][row_name]
            if row_type == "N":
                obj_rows.append(row_name)
            elif row_type == "E":
                eq_rows.append(row_name)
            elif row_type == "L":
                le_rows.append(row_name)
            elif row_type == "G":
                ge_rows.append(row_name)

        # Get variable bounds
        var_bounds = []
        for col_name in mps_data["col_order"]:
            if col_name in mps_data["bounds"]:
                bound_info = mps_data["bounds"][col_name]
                var_bounds.append((bound_info["lo"], bound_info["up"]))
            else:
                # Default: x >= 0 (standard LP convention)
                var_bounds.append((0.0, np.inf))

        # Count additional variables and constraints needed
        n_le = len(le_rows)
        n_ge = len(ge_rows)
        n_ineq = n_le + n_ge  # Number of slack variables needed

        # Count upper bound constraints (finite upper bounds)
        n_upper_bounds = sum(1 for lo, up in var_bounds if up < np.inf)

        # Total variables: original + slack for inequalities + slack for upper bounds
        n_slack_ineq = n_ineq
        n_slack_upper = n_upper_bounds
        n_total_vars = n_orig_cols + n_slack_ineq + n_slack_upper

        # Total constraints: equality + inequality (with slack) + upper bound constraints
        n_orig_eq = len(eq_rows)
        n_total_constraints = n_orig_eq + n_ineq + n_upper_bounds

        # Build objective vector c (extended with zeros for slack variables)
        c = np.zeros(n_total_vars)
        obj_name = mps_data["obj_name"]
        if obj_name:
            for col_name, coeffs in mps_data["columns"].items():
                if obj_name in coeffs:
                    c[col_idx[col_name]] = coeffs[obj_name]

        # Build unified constraint matrix A_eq and RHS b_eq
        A_eq = lil_matrix((n_total_constraints, n_total_vars))
        b_eq = np.zeros(n_total_constraints)

        row_idx = 0
        slack_idx = n_orig_cols  # Index for next slack variable

        # 1. Original equality constraints (E rows)
        for row_name in eq_rows:
            for col_name, coeffs in mps_data["columns"].items():
                if row_name in coeffs:
                    A_eq[row_idx, col_idx[col_name]] = coeffs[row_name]
            b_eq[row_idx] = mps_data["rhs"].get(row_name, 0.0)
            row_idx += 1

        # 2. Less-than-or-equal constraints (L rows): Ax + s = b, s >= 0
        for row_name in le_rows:
            for col_name, coeffs in mps_data["columns"].items():
                if row_name in coeffs:
                    A_eq[row_idx, col_idx[col_name]] = coeffs[row_name]
            # Add slack variable
            A_eq[row_idx, slack_idx] = 1.0
            b_eq[row_idx] = mps_data["rhs"].get(row_name, 0.0)
            row_idx += 1
            slack_idx += 1

        # 3. Greater-than-or-equal constraints (G rows): Ax - s = b, s >= 0
        for row_name in ge_rows:
            for col_name, coeffs in mps_data["columns"].items():
                if row_name in coeffs:
                    A_eq[row_idx, col_idx[col_name]] = coeffs[row_name]
            # Subtract slack variable
            A_eq[row_idx, slack_idx] = -1.0
            b_eq[row_idx] = mps_data["rhs"].get(row_name, 0.0)
            row_idx += 1
            slack_idx += 1

        # 4. Upper bound constraints: x + s_up = u, s_up >= 0
        for var_idx, (_lo, up) in enumerate(var_bounds):
            if up < np.inf:
                A_eq[row_idx, var_idx] = 1.0
                A_eq[row_idx, slack_idx] = 1.0
                b_eq[row_idx] = up
                row_idx += 1
                slack_idx += 1

        A_eq = A_eq.tocsr()

        # Handle lower bounds by shifting variables
        # For variables with lo != 0: x_new = x - lo, so A_eq stays same, b_eq adjusts
        c_shift = 0.0
        for var_idx, (lo, _up) in enumerate(var_bounds):
            if lo != 0 and lo > -np.inf:
                # x = x_new + lo, so c'x = c'x_new + c*lo
                c_shift += c[var_idx] * lo
                # Ax = A(x_new + lo) = Ax_new + A*lo, so b = b - A*lo
                b_eq -= A_eq[:, var_idx].toarray().flatten() * lo

        A_eq = A_eq.toarray()

        # Count free variables (variables with lo = -inf)
        n_free_vars = 0
        n_nonneg_vars = 0
        for lo, _up in var_bounds:
            if lo == -np.inf:
                n_free_vars += 1
            else:
                n_nonneg_vars += 1

        # Add slack variables to nonneg count
        n_nonneg_vars += n_slack_ineq + n_slack_upper

        # Create cone structure for SeDuMi format
        cone_structure = {"free_vars": n_free_vars, "nonneg_vars": n_nonneg_vars, "soc_cones": [], "sdp_cones": []}

        # Create metadata
        metadata = {
            "source": "MPS file",
            "format": "MPS (converted to SeDuMi)",
            "original_name": mps_data["name"],
            "objective_row": obj_name,
            "objective_shift": c_shift,
            "constraint_counts": {
                "original_equality": n_orig_eq,
                "original_less_equal": n_le,
                "original_greater_equal": n_ge,
                "upper_bound_constraints": n_upper_bounds,
                "total": n_total_constraints,
            },
            "original_dimensions": {"variables": n_orig_cols, "constraints": n_orig_eq + n_le + n_ge},
            "sedumi_dimensions": {"variables": n_total_vars, "constraints": n_total_constraints},
        }

        logger.info(
            f"Converted {problem_name}: LP problem "
            f"({n_orig_cols} original vars -> {n_total_vars} total vars, "
            f"{n_total_constraints} constraints)"
        )

        # Reshape for consistency with other loaders
        c = c.reshape(-1, 1)
        b_eq = b_eq.reshape(-1, 1)

        return ProblemData(
            name=problem_name,
            problem_class="LP",
            c=c,
            A_eq=A_eq,
            b_eq=b_eq,
            A_ub=None,
            b_ub=None,
            bounds=None,  # Bounds handled through SeDuMi cone structure
            cone_structure=cone_structure,
            metadata=metadata,
        )


if __name__ == "__main__":
    # Test script
    import sys

    if len(sys.argv) < 2:
        print("Usage: python mps_loader.py <path_to_mps_file>")
        sys.exit(1)

    file_path = sys.argv[1]

    try:
        loader = MPSLoader()
        problem = loader.load(file_path)
        print(f"Loaded problem: {problem}")
        print(f"  Variables: {problem._num_variables}")
        print(f"  Constraints: {problem._num_constraints}")
        print(f"  Bounds: {len(problem.bounds) if problem.bounds else 0} variables with bounds")

    except Exception as e:
        print(f"Error loading problem: {e}")
        import traceback

        traceback.print_exc()
        sys.exit(1)

File: scripts/data_loaders/matlab/mps_loader.m
function [A, b, c, K] = mps_loader(file_path)
    % Load NETLIB format .mps file and convert to SeDuMi format
    %
    % This function loads MPS (Mathematical Programming System) files and
    % converts them to standard SeDuMi format for use with MATLAB optimization
    % solvers. It mirrors the conversion performed by the Python loader
    % (scripts/data_loaders/python/mps_loader.py): inequality constraints are
    % converted to equality constraints via slack variables, and variable
    % bounds are handled through the cone structure.
    %
    % Unlike the Python loader, variables are reordered so that free variables
    % (K.f) come strictly before nonnegative variables (K.l), as required by
    % SeDuMi's cone convention.
    %
    % Input:
    %   file_path: Path to .mps file
    %
    % Output:
    %   A: Constraint matrix (sparse, m x n)
    %   b: Right-hand side vector (m x 1)
    %   c: Objective vector (n x 1)
    %   K: Cone structure (struct with fields K.f, K.l, K.q, K.s)
    %
    % Example:
    %   [A, b, c, K] = mps_loader('problems/NETLIB/mps_files/afiro.mps');

    try
        if nargin < 1
            error('mps_loader:InvalidInput', 'File path is required');
        end

        if ~ischar(file_path) && ~isstring(file_path)
            error('mps_loader:InvalidInput', 'File path must be a string or char array');
        end

        file_path = char(file_path);

        if ~exist(file_path, 'file')
            error('mps_loader:FileNotFound', 'File not found: %s', file_path);
        end

        fprintf('mps_loader: Parsing MPS file...\n');

        parsed = parse_mps_file(file_path);
        [A, b, c, K] = convert_to_sedumi(parsed);

        fprintf('mps_loader: Successfully loaded MPS problem\n');
        fprintf('  Variables (n): %d\n', size(A, 2));
        fprintf('  Constraints (m): %d\n', size(A, 1));
        fprintf('  Cone structure: K.f=%d, K.l=%d\n', K.f, K.l);

    catch ME
        fprintf('mps_loader: Error loading file %s\n', file_path);
        fprintf('Error: %s\n', ME.message);
        error('mps_loader:LoadFailed', 'Failed to load .mps file: %s\nOriginal error: %s', ...
            file_path, ME.message);
    end

end

function parsed = parse_mps_file(file_path)
    % Parse MPS format file into row/column/rhs/bounds structures.

    fid = fopen(file_path, 'r');
    if fid == -1
        error('mps_loader:FileOpenError', 'Cannot open file: %s', file_path);
    end

    row_order = {};
    row_type = containers.Map('KeyType', 'char', 'ValueType', 'char');
    row_index = containers.Map('KeyType', 'char', 'ValueType', 'double');
    obj_row_name = '';

    col_order = {};
    col_index = containers.Map('KeyType', 'char', 'ValueType', 'double');
    tri_i = [];
    tri_j = [];
    tri_v = [];

    rhs_map = containers.Map('KeyType', 'char', 'ValueType', 'double');
    bound_lo = containers.Map('KeyType', 'char', 'ValueType', 'double');
    bound_up = containers.Map('KeyType', 'char', 'ValueType', 'double');

    current_section = '';

    try
        while ~feof(fid)
            line = fgetl(fid);
            if ~ischar(line)
                continue;
            end
            raw_line = line;
            trimmed = strtrim(line);
            if isempty(trimmed) || trimmed(1) == '*'
                continue;
            end

            upper_line = upper(trimmed);
            if startsWith(upper_line, 'NAME')
                current_section = 'NAME';
                continue;
            elseif strcmp(upper_line, 'ROWS')
                current_section = 'ROWS';
                continue;
            elseif strcmp(upper_line, 'COLUMNS')
                current_section = 'COLUMNS';
                continue;
            elseif strcmp(upper_line, 'RHS')
                current_section = 'RHS';
                continue;
            elseif strcmp(upper_line, 'BOUNDS')
                current_section = 'BOUNDS';
                continue;
            elseif strcmp(upper_line, 'RANGES')
                current_section = 'RANGES';
                continue;
            elseif strcmp(upper_line, 'ENDATA')
                break;
            end

            switch current_section
                case 'ROWS'
                    parts = strsplit(raw_line);
                    parts = parts(~cellfun(@isempty, parts));
                    if numel(parts) >= 2
                        r_type = upper(parts{1});
                        r_name = parts{2};
                        row_order{end + 1} = r_name; %#ok<AGROW>
                        row_type(r_name) = r_type;
                        row_index(r_name) = numel(row_order);
                        if strcmp(r_type, 'N') && isempty(obj_row_name)
                            obj_row_name = r_name;
                        end
                    end

                case 'COLUMNS'
                    parts = strsplit(raw_line);
                    parts = parts(~cellfun(@isempty, parts));
                    if numel(parts) < 3
                        continue;
                    end
                    c_name = parts{1};
                    if ~isKey(col_index, c_name)
                        col_order{end + 1} = c_name; %#ok<AGROW>
                        col_index(c_name) = numel(col_order);
                    end
                    j = col_index(c_name);
                    k = 2;
                    while k <= numel(parts) - 1
                        r_name = parts{k};
                        val = str2double(parts{k + 1});
                        if isnan(val) || ~isKey(row_index, r_name)
                            break;
                        end
                        tri_i(end + 1) = row_index(r_name); %#ok<AGROW>
                        tri_j(end + 1) = j; %#ok<AGROW>
                        tri_v(end + 1) = val; %#ok<AGROW>
                        k = k + 2;
                    end

                case 'RHS'
                    parts = strsplit(raw_line);
                    parts = parts(~cellfun(@isempty, parts));
                    if numel(parts) < 3
                        continue;
                    end
                    k = 2;
                    while k <= numel(parts) - 1
                        r_name = parts{k};
                        val = str2double(parts{k + 1});
                        if isnan(val)
                            break;
                        end
                        rhs_map(r_name) = val;
                        k = k + 2;
                    end

                case 'BOUNDS'
                    parts = strsplit(raw_line);
                    parts = parts(~cellfun(@isempty, parts));
                    if numel(parts) < 3
                        continue;
                    end
                    b_type = upper(parts{1});
                    c_name = parts{3};
                    if ~isKey(bound_lo, c_name)
                        bound_lo(c_name) = 0.0;
                        bound_up(c_name) = Inf;
                    end
                    switch b_type
                        case 'LO'
                            bound_lo(c_name) = str2double(parts{4});
                        case 'UP'
                            bound_up(c_name) = str2double(parts{4});
                        case 'FX'
                            v = str2double(parts{4});
                            bound_lo(c_name) = v;
                            bound_up(c_name) = v;
                        case 'FR'
                            bound_lo(c_name) = -Inf;
                            bound_up(c_name) = Inf;
                        case 'MI'
                            bound_lo(c_name) = -Inf;
                        case 'PL'
                            bound_up(c_name) = Inf;
                        case 'BV'
                            bound_lo(c_name) = 0.0;
                            bound_up(c_name) = 1.0;
                    end

                case 'RANGES'
                    % RANGES constraints are not applied (parity with the
                    % Python loader, which parses but never uses them).
                    continue;
            end
        end

        fclose(fid);
    catch ME
        if fid ~= -1
            fclose(fid);
        end
        rethrow(ME);
    end

    parsed = struct();
    parsed.row_order = {row_order{:}}; %#ok<CCAT1>
    parsed.row_type = row_type;
    parsed.row_index = row_index;
    parsed.obj_row_name = obj_row_name;
    parsed.col_order = {col_order{:}}; %#ok<CCAT1>
    parsed.col_index = col_index;
    parsed.tri_i = tri_i;
    parsed.tri_j = tri_j;
    parsed.tri_v = tri_v;
    parsed.rhs_map = rhs_map;
    parsed.bound_lo = bound_lo;
    parsed.bound_up = bound_up;

end

function [A, b, c, K] = convert_to_sedumi(parsed)
    % Convert parsed MPS data to SeDuMi standard form:
    %   min  c'x
    %   s.t. A x = b, x in K
    %
    % Inequality constraints are converted to equalities with slack
    % variables; variable bounds are handled through the cone structure.
    % Variables are ordered [free vars][nonnegative vars] as required by
    % SeDuMi (nonnegative vars include original bounded vars and all slacks).

    n_orig_cols = numel(parsed.col_order);
    n_rows = numel(parsed.row_order);

    Coef = sparse(parsed.tri_i, parsed.tri_j, parsed.tri_v, n_rows, n_orig_cols);

    if ~isempty(parsed.obj_row_name)
        c_orig = full(Coef(parsed.row_index(parsed.obj_row_name), :))';
    else
        c_orig = zeros(n_orig_cols, 1);
    end

    eq_rows = {};
    le_rows = {};
    ge_rows = {};
    for i = 1:numel(parsed.row_order)
        r_name = parsed.row_order{i};
        r_type = parsed.row_type(r_name);
        if strcmp(r_type, 'E')
            eq_rows{end + 1} = r_name; %#ok<AGROW>
        elseif strcmp(r_type, 'L')
            le_rows{end + 1} = r_name; %#ok<AGROW>
        elseif strcmp(r_type, 'G')
            ge_rows{end + 1} = r_name; %#ok<AGROW>
        end
    end

    lo = zeros(n_orig_cols, 1);
    up = Inf(n_orig_cols, 1);
    for j = 1:n_orig_cols
        c_name = parsed.col_order{j};
        if isKey(parsed.bound_lo, c_name)
            lo(j) = parsed.bound_lo(c_name);
        end
        if isKey(parsed.bound_up, c_name)
            up(j) = parsed.bound_up(c_name);
        end
    end

    n_eq = numel(eq_rows);
    n_le = numel(le_rows);
    n_ge = numel(ge_rows);
    ub_var_indices = find(up < Inf)';
    n_ub = numel(ub_var_indices);

    n_slack = n_le + n_ge + n_ub;
    n_total_vars = n_orig_cols + n_slack;

    row_idx_of = @(name) parsed.row_index(name);
    eq_idx = cellfun(row_idx_of, eq_rows);
    le_idx = cellfun(row_idx_of, le_rows);
    ge_idx = cellfun(row_idx_of, ge_rows);

    A_eq_orig = Coef(eq_idx, :);
    A_le_orig = Coef(le_idx, :);
    A_ge_orig = Coef(ge_idx, :);

    A_eq_full = [A_eq_orig, sparse(n_eq, n_slack)];
    A_le_full = [A_le_orig, speye(n_le), sparse(n_le, n_ge), sparse(n_le, n_ub)];
    A_ge_full = [A_ge_orig, sparse(n_ge, n_le), -speye(n_ge), sparse(n_ge, n_ub)];

    if n_ub > 0
        A_ub_orig = sparse(1:n_ub, ub_var_indices, 1, n_ub, n_orig_cols);
        A_ub_full = [A_ub_orig, sparse(n_ub, n_le), sparse(n_ub, n_ge), speye(n_ub)];
        b_ub = up(ub_var_indices)';
    else
        A_ub_full = sparse(0, n_total_vars);
        b_ub = zeros(0, 1);
    end

    A_full = [A_eq_full; A_le_full; A_ge_full; A_ub_full];

    b_eq = zeros(n_eq, 1);
    for i = 1:n_eq
        if isKey(parsed.rhs_map, eq_rows{i})
            b_eq(i) = parsed.rhs_map(eq_rows{i});
        end
    end
    b_le = zeros(n_le, 1);
    for i = 1:n_le
        if isKey(parsed.rhs_map, le_rows{i})
            b_le(i) = parsed.rhs_map(le_rows{i});
        end
    end
    b_ge = zeros(n_ge, 1);
    for i = 1:n_ge
        if isKey(parsed.rhs_map, ge_rows{i})
            b_ge(i) = parsed.rhs_map(ge_rows{i});
        end
    end

    b_full = [b_eq; b_le; b_ge; b_ub(:)];
    c_full = [c_orig; zeros(n_slack, 1)];

    % Shift bounded variables with a nonzero finite lower bound so that
    % x_new = x - lo >= 0 (free variables and default lo=0 need no shift).
    for j = 1:n_orig_cols
        if lo(j) ~= 0 && lo(j) > -Inf
            b_full = b_full - A_full(:, j) * lo(j);
        end
    end

    % Reorder columns: free variables first, then nonnegative (original
    % bounded vars + all slacks), matching SeDuMi's K.f/K.l convention.
    free_mask = false(1, n_total_vars);
    free_mask(1:n_orig_cols) = (lo == -Inf)';
    perm = [find(free_mask), find(~free_mask)];

    A = A_full(:, perm);
    c = c_full(perm);
    b = b_full;

    K = struct();
    K.f = sum(free_mask);
    K.l = n_total_vars - K.f;
    K.q = [];
    K.s = [];

end
